### PR TITLE
Refactor

### LINKED ISSUE #16: Conform to the PEP 8 style guide
As title. 
There are some useful packages:
- [pycodestyle](https://github.com/PyCQA/pycodestyle).
- [autopep8](https://github.com/hhatto/autopep8).

### LINKED ISSUE #21: Fix the undefined `mode` in base_trainer.py
The `mode` in base_trainer.py is undefined. Please fix it.

### PR DESCRIPTION
**Pull request by peter850706**
- New features 
  - Add validation frequency (valid_freq) feature.
  - Support optional Automatic Mixed Precision (AMP) training.
  - Support [CyclicLR](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.CyclicLR) and others except [ReduceLROnPlateau](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau) because of the dynamic valid_freq.
  - Support gradient accumulation.
  - Support optional deterministic.
- Better abstractions (mainly for BaseTrainer and BasePredictor).
- Fix resuming training bugs
  - Save Monitor's state_dict instead of its instance to avoid pickle error.
  - Avoid non-deterministic behavior by saving torch_random_state and torch_cuda_random_state.
  - Avoid overwriting the current settings (e.g., saved_freq).
  - Foolproof design (lr_scheduler, monitor, amp).
- Miscellaneous
  - Conform PEP8 (Closes #16).
  - Remove np.random to avoid unexpected non-deterministic behavior.
  - Dataloader kwargs in config has 'train' and 'valid' section now.
  - Closes #21.

### CODE REVIEW

File: src/callbacks/monitor.py [hunk 1]
```diff
@@ -77,10 +80,19 @@ def state_dict(self):
...
+
+        if self.saved_freq != state_dict['saved_freq']:
+            LOGGER.warning(f"The saved_freq is changed from {state_dict['saved_freq']} to {self.saved_freq}.")
+
+        if self.early_stop != state_dict['early_stop']:
+            LOGGER.warning(f"The early_stop is changed from {state_dict['early_stop']} to {self.early_stop}.")
+
+        if self.not_improved_count >= self.early_stop:
+            LOGGER.critical(f"Load the checkpoint that should have to be early stopped.")
+            sys.exit()
```
Comment (vic85821) [inline src/callbacks/monitor.py:98]: Should we need to stop the process or just throw a warning message?

File: src/runner/trainers/base_trainer.py [hunk 2]
```diff
@@ -220,10 +221,21 @@ def load(self, path):
         checkpoint = torch.load(path, map_location='cpu')
         self.net.load_state_dict(checkpoint['net'])
         self.optimizer.load_state_dict(checkpoint['optimizer'])
-        if checkpoint['lr_scheduler'] is not None:
+
+        if self.lr_scheduler is not None and checkpoint['lr_scheduler'] is not None:
             self.lr_scheduler.load_state_dict(checkpoint['lr_scheduler'])
```
Comment (vic85821) [inline src/runner/trainers/base_trainer.py:260]: I'm not sure whether it fails when the lr scheduler is different from the one saved in the checkpoint.

**Review by vic85821 — COMMENTED** [review]
Reviewed.

File: src/data/dataloader.py [hunk 3]
```diff
@@ -25,6 +25,10 @@ def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None,
                          worker_init_fn=worker_init_fn)
 
         if dataset.type == 'train':
+            if self.drop_last and grad_accumulation_steps > len(self._index_sampler):
```
Comment (vic85821) [inline src/data/dataloader.py:28]: What would happen if the `drop_last = False` but `grad_accumulation_steps > len(self._index_sampler)`?

File: src/runner/predictors/base_predictor.py [hunk 4]
```diff
@@ -2,135 +2,72 @@
...
-        for metric_fn, metric in zip(self.metric_fns, metrics):
-            log[metric_fn.__class__.__name__] += metric.item() * batch_size
-
     def load(self, path):
         """Load the model checkpoint.
         Args:
             path (Path): The path to load the model checkpoint.
         """
-        checkpoint = torch.load(path, map_location=self.device)
+        checkpoint = torch.load(path, map_location='cpu')
```
Comment (vic85821) [inline src/runner/predictors/base_predictor.py:72]: Confused.

**Review by vic85821 — COMMENTED** [review]
Reviewed.

File: src/callbacks/monitor.py (hunk 1 above)
Comment (peter850706) [reply src/callbacks/monitor.py:98]: I think `sys.exit()` will stop the process.

File: src/runner/trainers/base_trainer.py (hunk 2 above)
Comment (peter850706) [reply src/runner/trainers/base_trainer.py:260]: Yep. As long as the `state_dict()` kwargs can be matched, it will not throw error messages. So do `self.net` and `self.optimizer`. I'm not sure how to prevent it and should we prevent it?

File: src/data/dataloader.py (hunk 3 above)
Comment (peter850706) [reply src/data/dataloader.py:28]: I think you mean `drop_last is True` but `grad_accumulation_steps > len(self._index_sampler)`.

If `drop_last is True`, the logic  is to drop the last uncompleted batch. Besides, the dataset can be divided by `batch_size`, which is defined as `len(self._index_sampler)`. 
Since we consider `grad_accumulation_steps * batch_size` as a completed batch, if `grad_accumulation_steps > len(self._index_sampler)`, the dataset (`len(self._index_sampler) * batch_size`) should be considered as **the last uncompleted batch**. That's why we need to throw an error.

File: src/runner/predictors/base_predictor.py (hunk 4 above)
Comment (peter850706) [reply src/runner/predictors/base_predictor.py:72]: Just for the consistency. Please refer to [L255 in base_trainer.py](https://github.com/cmlab-mira/template/blob/refactor/src/runner/trainers/base_trainer.py#L255). The reason is that `checkpoint['torch_random_state']` and `checkpoint['torch_cuda_random_state']` should be loaded as `torch.Tensor`, not `torch.CudaTensor`.

File: src/callbacks/monitor.py (hunk 1 above)
Comment (vic85821) [reply src/callbacks/monitor.py:98]: I see. But I mean whether it's necessary to stop the process. Maybe somebody needs to resume the training process after the `early stop` break?

File: src/data/dataloader.py (hunk 3 above)
Comment (vic85821) [reply src/data/dataloader.py:28]: No, I mean `drop_last = False` but `grad_accumulation_steps > len(self._index_sampler)`.

For example, the `grad_accumulation_steps` would be set to be larger than the total number of batch accidentally. In this case, it should raise an error in this place.

File: src/runner/predictors/base_predictor.py (hunk 4 above)
Comment (vic85821) [reply src/runner/predictors/base_predictor.py:72]: I see.

File: src/callbacks/monitor.py (hunk 1 above)
Comment (peter850706) [reply src/callbacks/monitor.py:98]: I think the logic of `early_stop` is that the users want to stop the training process (with certain setting) early. They could still resume the training process as `self.mode` or `self.target` has changed (`self.not_improved_count` would be reset to `np.inf`). But if not the case, I think we should prevent the users from resuming.

File: src/data/dataloader.py (hunk 3 above)
Comment (peter850706) [reply src/data/dataloader.py:28]: If `drop_last is False`, the logic is **not** to drop the last uncompleted batch. Therefore, the situation `grad_accumulation_steps > len(self._index_sampler)` can be considered as the whole dataset is an uncompleted batch.

**Comment by peter850706** [comment]
Close #21.
